docs/design/datacontracts/contract_csharp_api_design.cs

```diff
@@ -0,0 +1,386 @@
+        }
+
+        [DataContractGlobals("FeatureFlags", 2)]
+        public class FeatureFlags_2
+        {
+            public const int FeatureComInterop = 1;
+        }
+
+        [DataContractAlgorithm(1)]
+        class MethodTableContract_1 : ContractDefinitions.MethodTableContract, IAlgorithmContract
+        {
+            DataContracts.Target Target;
+            readonly uint ContractVersion;
+            public MethodTableContract_1(DataContracts.Target target, uint contractVersion) { Target = target; ContractVersion = contractVersion; }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { return new MethodTable(_target, methodTablePointer).dynamicTypeId; }
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return new MethodTable(_target, methodTablePointer).baseSizeAndFlags & 0x3FFFFFFF; }
+        }
+
+        // This is used for version 2 and 3 of the contract, where the dynamic type id is no longer present, and baseSize has a new limitation in that it can only be a value up to 0x1FFFFFFF in v3
+        [DataContractAlgorithm(2, 3)]
+        class MethodTableContract_2 : ContractDefinitions.MethodTableContract, IAlgorithmContract
+        {
+            DataContracts.Target Target;
+            readonly uint ContractVersion;
+            public MethodTableContract_2(DataContracts.Target target, uint contractVersion) { Target = target; }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer)
+            {
+                throw new NotImplementedException();
+            }
+            public virtual int BaseSize(TargetPointer methodTablePointer)
+            {
+                return new MethodTable(_target, methodTablePointer).baseSizeAndFlags & ((ContractVersion == 3) ? 0x1FFFFFFF : 0x3FFFFFFF);
+            }
+        }
+
+        // We use a source generator to generate the actual runtime properties, and api for working with the fields on this type.
+        //
+        // The source generator would fill in most of the apis, and provide a bunch of properties that give a granular failure model where if a particular field isn't defined, it fails at the access point
+        // This example shows access to a type.
+        [DataContractType]
+        partial struct MethodTable
+        {
+            partial void Get_dynamicTypeId_optional(ref int value);
+            partial void Get_baseSizeAndFlags(ref int value);
+
+            [DataContractLayout(1, 8)]
+            public class DataLayout1
+            {
+                [FieldOffset(0)]
+                public int dynamicTypeId;
+                [FieldOffset(4)]
+                public int baseSize;
+            }
+            [DataContractLayout(2, 4)]
+            public class DataLayout2
+            {
+                [FieldOffset(0)]
+                public int baseSize;
+            }
+
+            // The rest of this is generated by a source generator
```

Comment:
This auto-generated code doesn't seem needed for documentation. I think something like this would be a lot more concise and still convey all the info an implementation author needed:


## MethodTable version 1

```C#
struct MethodTable
{
    int dynamicTypeId; // offset 0
    int baseSize;      // offset 4
}
```

## MethodTable version 2

```C#
struct MethodTable
{
    int baseSize;      // offset 0
}
```

Reply:
Agreed... but this C# file was intended to express how the contract would be possibly be realized in C#. So, I needed to describe what the source generator might do to effectively discuss the apis available. As a consequence, I wrote out a bunch of boilerplate. I agree this isn't very obvious, and needs to be described more fully.